### PR TITLE
Fix injectivity radii for unitary matrix type manifolds

### LINKED ISSUE #573: Injectivity radius of a Rotation manifold
Hi, 

I was looking for the injectivity radius of rotation manifolds.  I found that you are always returning $\pi \sqrt{2}$:
```julia
injectivity_radius(Rotations(10))
> 4.442882938158366
```

Yet, when I played around this seems too tight as a bound. If I plug in a tangent vector with a norm of 5 the `log` of the `exp` still returns the original tangent vector:

```julia
rot = Rotations(5);
poi = random_point(rot);
tang = random_tangent(rot, poi);
norm(tang)
> 1.0
sum((4.5 * tang - log(rot, poi, exp(rot, poi, 4.5 * tang)))^2)
> -1.8797076257219422e-30
```

Does the example make sense? Is that the expected behavior? How /where did you find the injectivity radius for rotation manifolds?

Best, Constantin

### PR DESCRIPTION
**Pull request by kellertuer**
Fixes #573.

* [x] introduce new dispatch variants
* [x] update documentation
* [x] add tests
* [x] add a reference to an explaining blog or own section that derives the introduced/improved infectivity radii

### CODE REVIEW

File: src/manifolds/GeneralUnitaryMatrices.jl [hunk 1]
```diff
@@ -473,17 +473,48 @@ function get_vector_orthonormal!(
...
+    injectivity_radius(G::GeneraliUnitaryMatrices{n,𝔽,DeterminantOneMatrices})
 
-Return the injectivity radius on the [`Rotations`](@ref) and the [`OrthogonalMatrices`](@ref) `M`,
-which is globally
+Return the injectivity radius for general unitary matrix manifolds, where the determinant is $+1$,
+which is
 
 ````math
-    \operatorname{inj}_{\mathrm{O}(n)}(p) = \operatorname{inj}_{\mathrm{SO}(n)}(p) = π\sqrt{2}.
+    \operatorname{inj}_{\mathrm{SU}(n)}(p) = π.
```
Comment (sethaxen) [inline src/manifolds/GeneralUnitaryMatrices.jl:493]:
```suggestion
    \operatorname{inj}_{\mathrm{SU}(n)}(p) = π \sqrt{2}.
```

**Review by sethaxen — CHANGES_REQUESTED** [review]

File: src/manifolds/GeneralUnitaryMatrices.jl (hunk 1 above)
Comment (kellertuer) [reply src/manifolds/GeneralUnitaryMatrices.jl:493]: Oh, I was careful, but still I mixed up one in the docs.

**Comment by kellertuer** [comment]
I looked for literature and even discussed with a colleague; we could refer to Hüper, but since he uses the scaled metric compared to us, that even requires explanations then.

**Comment by sethaxen** [comment]
Should I write a short blog post that we can then link to? Or we could have a section of the docs devoted to miscellaneous notes where we could keep derivations like this and then link to them.

**Comment by kellertuer** [comment]
Hmpf. Sorry for so many commits, yesterday some dependencies broke, today it seems every line I write runs fine here but not on CI :/ but hopefully in a finite number of commits we might get epsilon-close to passing tests.

**Comment by sethaxen** [comment]
I wrote most of a blog post last night. Will try to either finish it tonight or adapt it into docs. It's ridiculous how much background one feels compelled to give whenever one writes about manifolds.

**Comment by kellertuer** [comment]
That sounds great! As I said for me both variants, a section in the docs or a blog post on your page are fine, what you prefer.

And yes, I know that feeling with the background ;)

**Comment by sethaxen** [comment]
I've posted a draft of the blog post at https://github.com/sethaxen/sethaxen.com/pull/1. Feel free to review if you'd like!

File: src/manifolds/GeneralUnitaryMatrices.jl [hunk 2]
```diff
@@ -487,17 +487,47 @@ function get_vector_orthonormal!(
...
 
 ````math
-    \operatorname{inj}_{\mathrm{O}(n)}(p) = \operatorname{inj}_{\mathrm{SO}(n)}(p) = π\sqrt{2}.
+    \operatorname{inj}_{\mathrm{U}(n)}(p) = π.
 ````
 """
+injectivity_radius(::GeneralUnitaryMatrices) = π
+
+@doc raw"""
+    injectivity_radius(G::GeneraliUnitaryMatrices{n,𝔽,DeterminantOneMatrices})
```
Comment (sethaxen) [inline src/manifolds/GeneralUnitaryMatrices.jl:501]:
```suggestion
    injectivity_radius(G::GeneralUnitaryMatrices{n,ℂ,DeterminantOneMatrices})
```

File: src/manifolds/GeneralUnitaryMatrices.jl [hunk 3]
```diff
@@ -487,17 +487,47 @@ function get_vector_orthonormal!(
...
+
+Return the injectivity radius for general unitary matrix manifolds, where the determinant is $+1$,
```
Comment (sethaxen) [inline src/manifolds/GeneralUnitaryMatrices.jl:503]:
```suggestion
Return the injectivity radius for general complex unitary matrix manifolds, where the determinant is $+1$,
```

File: src/manifolds/GeneralUnitaryMatrices.jl [hunk 4]
```diff
@@ -487,17 +487,47 @@ function get_vector_orthonormal!(
...
+    injectivity_radius(M::Rotations)
+    injectivity_radius(M::Rotations, ::ExponentialRetraction)
+
+Return the radius of injectivity on the [`Rotations`](@ref) manifold `M`, which is ``π\sqrt{2}``.
+"""
 injectivity_radius(::GeneralUnitaryMatrices{n,ℝ}) where {n} = π * sqrt(2.0)
+# same as above just resolving an ambiguity
+function injectivity_radius(::GeneralUnitaryMatrices{n,ℝ,DeterminantOneMatrices}) where {n}
+    return π * sqrt(2.0)
+end
```
Comment (sethaxen) [inline src/manifolds/GeneralUnitaryMatrices.jl:528]: No longer necessary
```suggestion
```

**Review by sethaxen — CHANGES_REQUESTED** [review]

File: src/manifolds/GeneralUnitaryMatrices.jl [hunk 5]
```diff
@@ -487,17 +487,47 @@ function get_vector_orthonormal!(
...
+
+```math
+    \operatorname{inj}_{\mathrm{SU}(n)}(p) = π \sqrt{2}.
+```
+"""
+function injectivity_radius(
+    ::GeneralUnitaryMatrices{n,𝔽,DeterminantOneMatrices},
+) where {n,𝔽}
+    return π * sqrt(2.0)
+end
```
[the remark this answers is not in the document]
Comment (kellertuer) [reply src/manifolds/GeneralUnitaryMatrices.jl:514]: Ah, well, we struggled already to give it a proper name earlier, maybe that is related ;)

**Comment by sethaxen** [comment]
Post is live at https://sethaxen.com/blog/2023/02/the-injectivity-radii-of-the-unitary-groups/

**Comment by kellertuer** [comment]
We could now either link to that in all the new docs or put that in a footnote and add the footnote link to all the docs – any preference?

**Comment by sethaxen** [comment]
Not really. Maybe footnote?

**Comment by kellertuer** [comment]
Ok, I will sketch that real quick (+- package dependencies in the docs being slow locally currently)

**Comment by kellertuer** [comment]
it is now a footnote, the only thing I do not like 100% is that literature references are also footnotes so the “1” is a bit lost between them.

Also we have a few footnote warnings still with-double footnotes we want to check but I am tired for today (fixed most refs that caused warnings)

**Comment by kellertuer** [comment]
There is one final warning which is a bit strange. For `TangentSpace` (defined in ManifoldsBase) we add a constructor in Manifolds (and import it correctly) but the doc strings of the constructor refer to `Manifolds.TangentSpaceAtPoint` and when generating that link, Documenter adds a manifolds/ too much (vector_bunlde is already in manifolds/) and hence the link fails the link check. I am not sure why. The rest should be fine.

**Comment by kellertuer** [comment]
And here's the link to the new doc strings rendered https://juliamanifolds.github.io/Manifolds.jl/previews/PR576/manifolds/generalunitary.html#ManifoldsBase.injectivity_radius-Tuple{Manifolds.GeneralUnitaryMatrices}

File: src/manifolds/GeneralUnitaryMatrices.jl [hunk 6]
```diff
@@ -487,17 +487,47 @@ function get_vector_orthonormal!(
...
+injectivity_radius(::GeneralUnitaryMatrices) = π
+
+@doc raw"""
+    injectivity_radius(G::GeneralUnitaryMatrices{n,ℂ,DeterminantOneMatrices})
+
+Return the injectivity radius for general complex unitary matrix manifolds, where the determinant is $+1$,
+which is[^1]
+
+```math
+    \operatorname{inj}_{\mathrm{SU}(n)}(p) = π \sqrt{2}.
```
Comment (sethaxen) [inline src/manifolds/GeneralUnitaryMatrices.jl:507]: Given that `p` is not an argument to the function, it makes sense to remove it.
```suggestion
    \operatorname{inj}_{\mathrm{SU}(n)} = π \sqrt{2}.
```

File: src/manifolds/GeneralUnitaryMatrices.jl [hunk 7]
```diff
@@ -487,17 +487,47 @@ function get_vector_orthonormal!(
...
-    injectivity_radius(M::Rotations)
+    injectivity_radius(G::GeneraliUnitaryMatrices)
 
-Return the injectivity radius on the [`Rotations`](@ref) and the [`OrthogonalMatrices`](@ref) `M`,
-which is globally
+Return the injectivity radius for general unitary matrix manifolds, which is[^1]
 
 ````math
-    \operatorname{inj}_{\mathrm{O}(n)}(p) = \operatorname{inj}_{\mathrm{SO}(n)}(p) = π\sqrt{2}.
+    \operatorname{inj}_{\mathrm{U}(n)}(p) = π.
```
Comment (sethaxen) [inline src/manifolds/GeneralUnitaryMatrices.jl:495]:
```suggestion
    \operatorname{inj}_{\mathrm{U}(n)} = π.
```

**Review by sethaxen — APPROVED** [review]
One minor suggestion. Otherwise, LGTM, thanks!

File: src/manifolds/GeneralUnitaryMatrices.jl (hunk 6 above)
Comment (kellertuer) [reply src/manifolds/GeneralUnitaryMatrices.jl:507]: Well spotted :)

**Pull request merged**